IPv6 only: use random MAC addresses

robmry opened the pull request:
**- What I did**

In a bridge network with IPv4 enabled, when no MAC address is explicitly configured for a container, the MAC address was generated from the IPv4 address. So, when the container was stopped and its IP addresses were quickly recycled for another container, ARP/neighbour caches were still correct because there was a fixed relationship between MAC and IP addresses.

It's not possible to generate a unique MAC address from an IPv6 address (at least, not with a prefix length of fewer than 80 bits, and the prefix length should normally be at-least 64-bits).

So, for a container in an IPv6-only network, a new scheme is needed or when an IPv6 address gets recycled for another container the neighbour cache will contain the wrong mapping from that IPv6 address to a MAC address.

- Also, fixes https://github.com/moby/moby/issues/48082
  - An issue with Portainer setting a MAC address but no IP address when restarting containers, resulting in duplicate MAC addresses derived from recycled IP addresses. Now, the old random MAC address won't be recycled along with the IP address.

**- How I did it**

- Abandoned the fixed mapping from IPv4 address to MAC address.
  - Now, container interfaces on a bridge network always have a randomly generated MAC address.
  - Send an unsolicited ARP or Neighbour Advertisement when bringing up a new interface, to update caches.

Some of this was written as prototype by @corhere, I've picked up some changes from [this branch](https://github.com/moby/moby/compare/master...corhere:moby:libn/gratuitous-arp). However, setting sysctl `arp_notify` to have the kernel send the unsolicited ARP/NA messages didn't work out ...
- the kernel didn't send an ARP/NA when the interface came up
  - https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1989187
- as a workaround, we tried triggering the ARP/NA by setting the interface's MAC address
  - that was fine for a single initial message, but
  - setting the MAC address again later to trigger ARP/NA resends, flushed entries from the ARP/neighbour caches.

So, the ARP/NA messages are now sent from user space.

The number of unsolicited ARP/NA messages, and the interval between them, are configurable:
- The default is to send 3 messages at 1s intervals.
- That can be overridden in "docker network create" using:
    `-o com.docker.network.advertise_addr_nmsgs=3`
    `-o com.docker.network.advertise_addr_ms=1000`
- Or, in daemon.json for each driver:
    ```
        "default-network-opts": {
          "bridge": {
            "com.docker.network.advertise_addr_nmsgs": "3",
            "com.docker.network.advertise_addr_ms": "1000"
          }
        }
    ```
- The allowed range is 0-3 for count, and 100-2000ms for the interval.
- Setting count to 0 disables the unsolicited ARP/NA messages.
- The default bridge will always use the built-in defaults, it cannot be configured.

Note that random MAC addresses are now used on the default bridge network.
- Until now, as a special case for the default bridge network, IPv6 addresses have been based on the MAC address rather than being IPAM-assigned. So, as the MAC address was based on the IPv4 address unless it was specifically configured, there was a fixed mapping from IPv4 to MAC to IPv6 address. That no longer holds. But, all three addresses can be explicitly configured.

**- How to verify it**

New integration test to check that ARP/NA messages are sent.

Also tested while running `iperf` ...
- In a moby dev container, running in DD on an M2 Macbook.
- `iperf -s` in one container on a bridge, and `iperf -c <addr>` on another -> ~115Gbps TCP
- In one of the containers, pinged a third container's v4 and v6 addresses, to populate the neighbour cache.
  - Restarted the third container a few times, with the same IP address.
  - Used `ip neigh show` to check the iperf container's neighbour cache was immediately updated each time.
- Repeated that, but pinging from the other iperf container.
- Just to be sure, checked that with `nmsgs=0`, the neighbour cache wasn't updated when the third container restarted.

To check for goroutine leaks ...
- Started a bunch of containers running a 4s sleep (so that all three ARP and NA messages would be sent)...
  - `i=0; while [ $i -lt 500 ]; do ( docker run --rm -d --network b46 busybox sleep 4 & ); i=$((i+1)); done`
- Then the same again, but with a 1s sleep so that the ARP/NA goroutine would be cancelled.
- `docker info` reported that it got back to 49 goroutines.

**- Description for the changelog**
```markdown changelog
- Container interfaces in bridge networks now use randomly generated MAC addresses.
  - Unsolicited ARP / Neighbour Advertisement messages will be sent when the interfaces are started
    so that, when IP addresses are reused, they're associated with the newly generated MAC address.
  - IPv6 addresses in the default bridge network are now IPAM-assigned, rather than being derived
    from the MAC address.
  - Driver option "com.docker.network.advertise_addr_nmsgs" can be used to set the number of
    unsolicited ARP and NA packets sent for each address, from 0 to 3, the default is 3.
  - Driver option "com.docker.network.advertise_addr_ms" can be used to set the interval between
    unsolicited ARP and NA packets sent for each address, from 100 to 2000ms, the default is 1000ms.
```

corhere committed:
libn/netutils: put more entropy into random MACs

With randomly-generated MAC addresses going into wider use in the most
commonly used container network drivers, it is imperative to avoid
address collisions. Increase the number of random bits in randomly-
generated MAC addresses from 32 bits to 46 bits.

Signed-off-by: Cory Snider <csnider@mirantis.com>

robmry committed:
bridge/macvlan endpoints always use a random MAC address

Now a gratuitous/unsolicted ARP is sent, there's no need to
use an IPv4-based MAC address to preserve arp-cache mappings
between an endpoint's IP addresses and its MAC addresses.

Because a random MAC address is used for the default bridge,
it no longer makes sense to derive container IPv6 addresses
from the MAC address. This "postIPv6" behaviour was needed
before IPv6 addresses could be configured, but not now. So,
IPv6 addresses will now be IPAM-allocated on the default
bridge network, just as they are for user-defined bridges.

Signed-off-by: Rob Murray <rob.murray@docker.com>

akerouanton commented on libnetwork/osl/interface_linux.go:
```diff
@@ -276,19 +288,104 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+			log.G(ctx).WithFields(log.Fields{
+				"ifi":   ifIndex,
+				"error": err,
+			}).Debug("netlink error while waiting for interface up")
```
As discussed on Slack, I saw this error while testing this PR (second-to-last log):

```
DEBU[2024-11-06T10:11:28.907169337Z] link has been set to up                       spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
DEBU[2024-11-06T10:11:28.907424671Z] link update                                   flags="deviceFlags(IFF_UP | IFF_BROADCAST | IFF_MULTICAST | IFF_LOWER_UP)" iface=eth0 ifi=14 spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
DEBU[2024-11-06T10:11:28.907488212Z] link update                                   flags="deviceFlags(IFF_UP | IFF_BROADCAST | IFF_MULTICAST | IFF_LOWER_UP)" iface=eth0 ifi=14 spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
DEBU[2024-11-06T10:11:28.907534337Z] link update                                   flags="deviceFlags(IFF_UP | IFF_BROADCAST | IFF_RUNNING | IFF_MULTICAST | IFF_LOWER_UP)" iface=eth0 ifi=14 spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
DEBU[2024-11-06T10:11:28.907587671Z] netlink error while waiting for interface up  error="Receive failed: resource temporarily unavailable" ifi=14 spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
DEBU[2024-11-06T10:11:28.907681129Z] applying gratuitous ARP workaround            iface=eth0 ifi=14 ip=172.19.0.2/16 ip6="fd9b:a045:3ddd::2/64" mac="02:42:ac:13:00:02" spanID=ec4fd13878d1114f traceID=e489e9906a34cf8491e39f49a60d219b
```

After investigating, we found out that this is totally benign. What happens is that the `defer close(upped)` below is used to signal to `LinkSubscribeWithOptions` it should stop watch the interface. But then, in vishvananda/netlink these two goroutines are racing against each other:

- https://github.com/vishvananda/netlink/blob/main/link_linux.go#L2435-L2438
- https://github.com/vishvananda/netlink/blob/main/link_linux.go#L2453-L2459

That second goroutine doesn't special-case errors induced by the `s.Close()` call made by the 1st goroutine.

So either we need to fix vishvananda (IMO sending this EAGAIN is really fishy!), or we need to add another defer running before we close the channel to signal to `ErrorCallback` that it shouldn't log anything anymore.

robmry replied:
That's done - thank you ... I also added a check for the `update` channel closing early, in case the netlink socket read fails for some other reason, causing the function in the netlink package to bail out before the `done` channel is closed.

akerouanton commented on libnetwork/network.go:
```diff
@@ -455,7 +454,6 @@ func (n *Network) CopyTo(o datastore.KVObject) error {
 	dstN.enableIPv4 = n.enableIPv4
 	dstN.enableIPv6 = n.enableIPv6
 	dstN.persist = n.persist
-	dstN.postIPv6 = n.postIPv6
```
I'm wondering if this could cause issues when downgrading? Maybe we should keep this field for one more major release (or until the latest Mirantis release to support that behavior reaches EOL)?

robmry replied:
It works out ok, because the default bridge gets set up again by the old build.

(We don't support `live-restore` over a major version, but that happens to be fine too. Running containers keep their addresses on upgrade anyway, containers created after upgrade get IPAM addresses, and keep them on live-downgrade.)

akerouanton reviewed (commented):
Beside my comment about downgrades, and the EAGAIN log, LGTM.

robmry committed:
Configurable count and interval for gratuitous ARP/NA messages

The default is to send 3 messages at 1s intervals.

That can be overridden in "docker network create" using:
  -o com.docker.network.advertise_addr_nmsgs=3
  -o com.docker.network.advertise_addr_ms=1000

Or, in daemon.json for each driver:
  "default-network-opts": {
    "bridge": {
      "com.docker.network.advertise_addr_nmsgs": "3",
      "com.docker.network.advertise_addr_ms": "1000"
    }
  }

The allowed range is 0-3 for the number of messages, and
100-2000ms for the interval. Setting nmsgs to 0 disables the
gratuitous ARP/NA messages.

The default bridge will always use the built-in defaults,
it cannot be configured.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
> In the kernel, the `RTM_SETLINK` call to update the MAC address, to trigger ARP/NA transmission, ends up in [dev_set_mac_address()](https://github.com/torvalds/linux/blob/master/net/core/dev.c#L9087). It won't actually set the address unless it's changed, it'll just call the change notifiers. So, it shouldn't be disruptive.

Hmm ... setting the MAC address clears the neighbour cache, which isn't so good.

robmry committed:
Test unsolicited ARP/NA on interface creation

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Send unsolicited ARP/NA requests when bringing up interfaces

Co-authored-by: Cory Snider <csnider@mirantis.com>
Co-authored-by: Rob Murray <rob.murray@docker.com>
Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
> > In the kernel, the `RTM_SETLINK` call to update the MAC address, to trigger ARP/NA transmission, ends up in [dev_set_mac_address()](https://github.com/torvalds/linux/blob/master/net/core/dev.c#L9087). It won't actually set the address unless it's changed, it'll just call the change notifiers. So, it shouldn't be disruptive.
> 
> Hmm ... setting the MAC address clears the neighbour cache, which isn't so good.

Switched to sending the ARP/NA packets from user space.

akerouanton commented on libnetwork/osl/interface_linux.go:
```diff
@@ -499,6 +538,29 @@ func (n *Namespace) prepAdvertiseAddrs(ctx context.Context, i *Interface, ifInde
+// ValidateAdvertiseAddrNMsgs returns an error if count is not within the allowed
+// range for the number of unsolicited ARP/NA messages sent to advertise an
+// interface's addresses when it is configured.
+func ValidateAdvertiseAddrNMsgs(count int) error {
```
The `osl` pkg isn't interested in validating user input. Could you move both funcs into `libnetwork` please?

robmry replied:
The user-input validation [is in libnetwork](https://github.com/robmry/moby/blob/68a7767ae0ef9b20d7e8dcccca5221a989d7a7b6/libnetwork/network.go#L546-L559) - but, once it's got as far as having an integer, it asks `osl` to range-check it. The `osl` package sends the ARP/NA messages, so I think it makes sense for the rules about what it's willing to send to live there too?

robmry replied:
I've done that - and added range checks to `osl`'s option setters.

akerouanton commented on libnetwork/osl/interface_linux.go:
```diff
@@ -69,7 +95,13 @@ type Interface struct {
 	sysctls     []string
-	ns          *Namespace
+	// advertiseAddrNMsgs is the number of unsolicited ARP/NA messages that will be sent to
+	// advertise the interface's addresses.
```
Could you please mention that unsolicited ARP/NA are disabled when `advertiseAddrNmsgs` is set to zero?

robmry replied:
Done.

akerouanton reviewed (commented):
I still need to take a look at the last commit, but here's a first batch of comments.

robmry commented on libnetwork/osl/interface_linux.go:
```diff
@@ -276,22 +292,217 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+	// Send the rest in the background.
+	go func() {
+		defer cleanup()
+		ctx, span := otel.Tracer("").Start(context.WithoutCancel(ctx), "libnetwork.osl.advertiseAddrs")
```
Thank you for helping me get the OTEL machine working! As discussed - the trace looks ok .. it hangs off `daemon.containerStart` -> ... -> `libnetwork.osl.AddInterface`, and the 2s duration of span `libnetwork.osl.advertiseAddrs` isn't included in the duration of its parents.

![Screenshot 2024-12-10 at 11 40 08](https://github.com/user-attachments/assets/dd621f87-17c5-4141-9e86-b4992df56c79)

robmry commented on libnetwork/internal/l2disco/unsol_arp_linux.go:
```diff
@@ -0,0 +1,75 @@
+	"fmt"
+	"net"
+	"slices"
+	"syscall"
```
Done.

robmry commented on libnetwork/internal/l2disco/unsol_arp_linux.go:
```diff
@@ -0,0 +1,75 @@
+
+// To run in container netns.
+func NewUnsolARP(ip net.IP, mac net.HardwareAddr, ifIndex int) (*UnsolARP, error) {
+	sd, err := syscall.Socket(syscall.AF_PACKET, syscall.SOCK_DGRAM, int(htons(syscall.ETH_P_ARP)))
```
Done.

robmry commented on libnetwork/internal/l2disco/unsol_na_linux.go:
```diff
@@ -0,0 +1,74 @@
+
+// https://datatracker.ietf.org/doc/html/rfc4861#section-4.4
+func NewUnsolNA(ip net.IP, mac net.HardwareAddr, ifIndex int) (*UnsolNA, error) {
+	sd, err := syscall.Socket(syscall.AF_INET6, syscall.SOCK_RAW, syscall.IPPROTO_ICMPV6)
```
Done.

akerouanton commented on libnetwork/internal/l2disco/unsol_arp_linux.go:
```diff
@@ -0,0 +1,80 @@
+// NewUnsolARP returns a pointer to an object that can send unsolicited ARPs on
+// the interface with ifIndex, for ip and mac.
+func NewUnsolARP(ip net.IP, mac net.HardwareAddr, ifIndex int) (*UnsolARP, error) {
+	sd, err := unix.Socket(unix.AF_PACKET, unix.SOCK_DGRAM, int(htons(unix.ETH_P_ARP)))
```
I'm the one telling you we need to call `CloseOnExec` below, but I now realize we should instead `|` the 3rd argument with `SOCK_CLOEXEC`.

[`man 2 socket` says](https://man7.org/linux/man-pages/man2/socket.2.html#:~:text=Since%20Linux%202.6.27%2C%20the%20type%20argument%20serves%20a%20second%20purpose%3A%20in%0A%20%20%20%20%20%20%20addition%20to%20specifying%20a%20socket%20type%2C%20it%20may%20include%20the%20bitwise%0A%20%20%20%20%20%20%20OR%20of%20any%20of%20the%20following%20values%2C%20to%20modify%20the%20behavior%20of%0A%20%20%20%20%20%20%20socket()%3A):

       Since Linux 2.6.27, the type argument serves a second purpose: in
       addition to specifying a socket type, it may include the bitwise
       OR of any of the following values, to modify the behavior of
       socket():

robmry replied:
Doh, I guess I've spent too long away from C! Thank you.

akerouanton commented on libnetwork/internal/l2disco/unsol_na_linux.go:
```diff
@@ -0,0 +1,80 @@
+// Advertisements on the interface with ifIndex, for ip and mac.
+// https://datatracker.ietf.org/doc/html/rfc4861#section-4.4
+func NewUnsolNA(ip net.IP, mac net.HardwareAddr, ifIndex int) (*UnsolNA, error) {
+	sd, err := unix.Socket(unix.AF_INET6, unix.SOCK_RAW, unix.IPPROTO_ICMPV6)
```
We need to do the same here.

robmry replied:
Done (and the ones in the test code for good measure).

akerouanton reviewed (commented):
I'm still taking another look but posting my comments around `SOCK_CLOEXEC` early.

akerouanton commented on libnetwork/osl/interface_linux.go:
```diff
@@ -12,6 +13,7 @@ import (
+	"github.com/docker/docker/libnetwork/internal/l2disco"
 	"github.com/docker/docker/libnetwork/ns"
 	"github.com/docker/docker/libnetwork/types"
 	"github.com/pkg/errors"
```
We should remove this import in a follow-up, such that we don't need to alias the stdlib `errors` pkg.

thaJeztah replied:
Oh!  LOL, missed this comment; just wrote up the same above; I guess I can open a PR already to get rid of it

thaJeztah replied:
☝️ opened https://github.com/moby/moby/pull/49224 for consideration

akerouanton commented on libnetwork/osl/interface_linux.go:
```diff
@@ -276,22 +292,217 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+
+	// Send an initial message. If it fails, skip the resends.
+	if send(ctx, 0) != nil {
+		return
```
If the initial unsolicited ARP/NA messages can't be sent, should we log the error, or even bail out?

robmry replied:
It'll be logged by `send` itself ... but I figured it'd be best not to treat errors here as fatal - just in case there's some host/config where it never works, although that seems unlikely. (If it fails, the container will be up and running, and lack of ARP/NA won't always be an issue.)

But, now I realise setting the network's `advertise_addr_nmsgs` to zero would be a workaround if things do go wrong. So, yes - it returns the error now. Thank you.

robmry replied:
That found a bug ... Swarm doesn't set a MAC address in its `Interface`, so it wasn't sending anything.

Fixed by reading the address from the interface in that case, and checked that ARPs are now sent when starting a swarm service.

robmry replied:
Well, that was confusing! Not-sending an ARP/NA when the `Interface` doesn't have one was the right thing after all - in the swarm case, it was a bridge with no addresses moving netns and for its bridge/overlay networks, the ARPs were fine. L3-ipvlan endpoints also have no MAC address to advertise. So, just bail out when no MAC address is set up.

robmry replied:
Urgh, had to make another update ... I remembered `waitForIfUpped` waits for the interface to be `UP`+`RUNNING`+`LOWER_UP` - but, if it doesn't get there within 5s, container startup just carries on. (Before this PR, it didn't try to wait for the interface to report "up".)

Treating failure to send an ARP/NA as an error, rather than just logging, would unintentionally have made a "down" interface into an error.

So ... I got rid of the goroutine that waited for the interface to come up, made `waitForIfUpped` just block, and return a flag to say whether the interface is ready for unsolicited ARPs/NAs.

(The goroutine was there so to make it possible to subscribe to netlink updates before setting the interface "up". But, that hasn't been needed since I added `ListExisting` to the netlink options.)

That in turn made me realise the `NO_CARRIER` check was nonsense ... wrong flag! I got rid of it, but perhaps should bail out early if `LOWER_UP` is unset instead.

robmry committed:
Treat failure to send an initial ARP/NA as an error

If this causes problems, setting nmsgs=0 is a workaround.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
Switching from logging to erroring when an ARP/NA fails was a bit traumatic! It was getting lucky with a few things before, so ARP/NA sends were attempted when it wasn't possible or necessary. But I think it's right now, and a bit simpler.

The changes are all in `interface_linux.go`:
- handle an `Interface` with no configured MAC address (l3-vlans, swarm bridges, ...)
- just wait for the interface to be `UP`, ignore `LOWER_UP` and `RUNNING`
- removed the check for `NO_CARRIER`, because it was just wrong
- got rid of the goroutine that waited for the interface to come up, it can just block now
- only try to send an ARP/NA if the interface did come up

akerouanton reviewed (approved):
LGTM

vvoland commented on libnetwork/osl/interface_linux.go:
```diff
@@ -276,22 +323,232 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+
+	timer := time.NewTimer(5 * time.Second)
+	defer timer.Stop()
+	defer close(upped)
```
Is there a chance that 
https://github.com/moby/moby/blob/f95e4b3d22260eb3322b4be13d391b8db2d4199f/vendor/github.com/vishvananda/netlink/link_linux.go#L2452

won't return immediately after `s.Close()`?

The goroutine inside `linkSubscribeAt` closes the `update` channel before it exists. 
Wondering if we should wait for the `update` after requesting the `linkSubscribeAt` to end, so that we're sure that the goroutine actually exited?

```suggestion
	defer func() {
		close(upped)
		select {
		  case <-update:
		  case time.After(time.Second * 3):
		     // log warning/error
		}
	}
```

robmry replied:
I've done that ... slightly modified because I think another message on the `update` could race the close - so the select would still terminate before the netlink goroutine terminates and closes the channel.

vvoland commented on libnetwork/osl/interface_linux.go:
```diff
@@ -276,22 +323,232 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+
+	for {
+		select {
+		case <-timer.C:
```
```suggestion
		case <-time.After(time.Second * 5)
```

nit: With Go 1.23, there's no performance advantage of `time.NewTimer` over `time.After`, so we could just use the latter since it's a bit simpler.

robmry replied:
I've changed it to `time.After` - but created before the for-loop, so it doesn't create a new timer each time a message arrives on the `update` channel.

vvoland commented on libnetwork/drivers/bridge/bridge_linux.go:
```diff
@@ -1261,33 +1252,10 @@ func (d *driver) CreateEndpoint(ctx context.Context, nid, eid string, ifInfo dri
+	log.G(ctx).WithFields(log.Fields{
+		"hostifname": host.Attrs().Name,
+		"ifi":        host.Attrs().Index,
+	}).Debugf("bridge endpoint host link is up")
```
nit: don't bother if not doing other changes

```suggestion
	}).Debug("bridge endpoint host link is up")
```

robmry replied:
Done.

vvoland commented on integration/internal/network/l2disco_linux.go:
```diff
@@ -0,0 +1,210 @@
+		defer unix.Close(sd)
+		for {
+			buf := make([]byte, 50)
+			n, err := unix.Read(sd, buf)
+			select {
+			case <-stopC:
+				return
+			default:
+			}
```
Shouldn't we check for the stop *before* actually doing the Read syscall? It's probably more expensive than the rest of the loop (checking error and appending to the slice).

```suggestion
			select {
			case <-stopC:
				return
			default:
			}
			n, err := unix.Read(sd, buf)
```

robmry replied:
Ahh, right... it's not clear, so I've added some comments to the function that might help.

It's a blocking `Read` with a timeout ... so, if no packet is received for one second it'll return an `EWOULDBBLOCK`.

When that happens, there's a good chance the caller has asked for packet collection to stop (by calling the function returned by `collectPackets`). In that case, this packet-collecting goroutine just needs to exit ... there's probably no packet to process and append to `pkts` - but it doesn't matter, that packet's not needed.

So, checking `stopC` before the read would be more-or-less equivalent - because the `EWOULDBLOCK` would just cause a `continue`, back to the top of the loop where, `stopC` would be checked ... but there's no need to go back around the loop.

robmry commented on libnetwork/osl/interface_linux.go:
```diff
@@ -276,22 +323,232 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+// Note that the kernel's arp_notify sysctl setting is not respected.
+func (n *Namespace) advertiseAddrs(ctx context.Context, ifIndex int, i *Interface, nlh nlwrap.Handle) error {
+	mac := i.MacAddress()
+	logFields := log.Fields{
+		"iface": i.dstName,
+		"ifi":   ifIndex,
+		"mac":   mac.String(),
+		"ip4":   i.Address(),
+		"ip6":   i.AddressIPv6(),
+	}
+	ctx = log.WithLogger(ctx, log.G(ctx).WithFields(logFields))
```
Done.

corhere commented on libnetwork/osl/interface_linux.go:
```diff
@@ -276,22 +323,246 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+		}
+	}()
+
+	send := func(ctx context.Context, nSent int) error {
```
The send function doesn't really care about the `nSent` parameter. It's just there to pass through logging metadata. It smells to me like the advertisement number is nothing more than context for the function call.
```go
send(log.WithLogger(ctx, log.G(ctx).WithFields(log.Fields{"n": nSent + 1})))
```

robmry replied:
Ok.

corhere commented on libnetwork/osl/interface_linux.go:
```diff
@@ -276,22 +323,246 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+		defer span.End()
+		ticker := time.NewTicker(i.advertiseAddrInterval)
+		defer ticker.Stop()
+		for nSent, nToSend := 1, i.advertiseAddrNMsgs; nSent < nToSend; nSent += 1 {
```
Why are you copying `i.advertiseAddrNMsgs` into a local variable? If it's some ploy to prevent data races, I don't think it's effective.

Range-over-int can be used here; it's supported as of Go 1.22.
```suggestion
		for n := range i.advertiseAddrNMsgs - 1 {
```

robmry replied:
No ploy, it had that shape in the prev commit, before the config existed. But, sure.

robmry commented on libnetwork/internal/l2disco/unsol_arp_linux.go:
```diff
@@ -0,0 +1,79 @@
+// NewUnsolARP returns a pointer to an object that can send unsolicited ARPs on
+// the interface with ifIndex, for ip and mac.
+func NewUnsolARP(ip net.IP, mac net.HardwareAddr, ifIndex int) (*UnsolARP, error) {
+	sd, err := unix.Socket(unix.AF_PACKET, unix.SOCK_DGRAM|unix.SOCK_CLOEXEC, int(htons(unix.ETH_P_ARP)))
```
Ok.

robmry commented on libnetwork/internal/l2disco/unsol_arp_linux.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+// Send sends an unsolicited ARP message.
+func (ua *UnsolARP) Send() error {
```
Because it makes the socket's lifecycle very simple ... create, send messages, close.

With a socket per namespace, either the socket has to be created and kept around until the container's deleted, or it needs to be closed when ARPs have been sent for all interfaces (then another socket needs to be opened if the container is connected to another network). It also means the `Namespace` has some extra state to track.

I'm not convinced it's worthwhile - so, I've put it into a separate commit. Can squash it, remove, or fiddle around some more.

corhere replied:
The per-namespace socket's lifecycle can be just as simple: create in `osl.NewSandbox()`, send messages, close in `(*osl.Namespace).Destroy()`. It's what we currently do for the netlink socket. And I'm going to need a socket which the overlay driver can use to send unsolicited proxy ARP and ND advertisements for an endpoint of a container _on another host_. One persistent socket per local namespace is going to scale much more predictably than opening and closing ephemeral sockets every time a container anywhere in the Swarm attaches to an overlay network.

robmry replied:
As soon as the socket is shared, it needs to be initialised and stored somewhere, restored on live-restore, and not closed until anything that's using it stops (or errors caused by that race need special handling). Keeping it open until the namespace is deleted doesn't help with that.

So, I think we should just keep the "WIP" commit as-is (squashed into the prev commit).

corhere replied:
What am I missing? What makes the ARP socket fundamentally different from the netlink socket field of the Namespace struct? Why would it be inappropriate for them to have the same lifetime?

corhere replied:
As discussed offline, we have decided to drop the WIP commit. The initial implementation with ephemeral sockets is a local optimum for the implemented use case.

robmry replied:
It's gone.

robmry commented on libnetwork/internal/l2disco/unsol_arp_linux.go:
```diff
@@ -0,0 +1,79 @@
+// NewUnsolARP returns a pointer to an object that can send unsolicited ARPs on
+// the interface with ifIndex, for ip and mac.
+func NewUnsolARP(ip net.IP, mac net.HardwareAddr, ifIndex int) (*UnsolARP, error) {
+	sd, err := unix.Socket(unix.AF_PACKET, unix.SOCK_DGRAM|unix.SOCK_CLOEXEC, int(htons(unix.ETH_P_ARP)))
```
I considered using github.com/mdlayher/socket, but [it should not be used directly in end user applications](https://github.com/mdlayher/socket#:~:text=This%20package%20should%20not%20be%20used%20directly%20in%20end%20user%20applications.).

A system under so much pressure that it can't squeeze out an ARP packet for any significant period of time has bigger issues than this thread.

(github.com/mdlayher/arp has no tagged releases, and no updates for a couple of years. I didn't look any closer than that.)

robmry commented on libnetwork/internal/l2disco/unsol_arp_linux.go:
```diff
@@ -0,0 +1,79 @@
+		_ = unix.Close(ua.sd)
+		ua.sd = -1
+	}
+}
```
Ok.

robmry commented on libnetwork/internal/l2disco/unsol_na_linux.go:
```diff
@@ -0,0 +1,79 @@
+// NewUnsolNA returns a pointer to an object that can send unsolicited Neighbour
+// Advertisements on the interface with ifIndex, for ip and mac.
+// https://datatracker.ietf.org/doc/html/rfc4861#section-4.4
+func NewUnsolNA(ip net.IP, mac net.HardwareAddr, ifIndex int) (*UnsolNA, error) {
```
Removed.

robmry replied:
It's back again, because it's used now.

robmry commented on libnetwork/internal/l2disco/unsol_na_linux.go:
```diff
@@ -0,0 +1,79 @@
+	if err := unix.Bind(sd, saBind); err != nil {
+		_ = unix.Close(sd)
+		return nil, fmt.Errorf("bind socket: %w", err)
+	}
```
Ok, I've had a go ... it's a bit weird having to bind to an address to get a `net.PacketConn` to feed to the `ipv6.PacketConn`, but probably ok. Maybe I've missed a better way to do it? I don't think anything's leaking, but it's much harder to be sure.

robmry replied:
Hopefully it'll be ok now ... the problem was when the interface only had link-local IPv6 addresses, and the fix was to specify a`IfIndex` in the `ipv6.ControlMessage`. I've added a specific integration test to cover NA messages for LL addresses.

robmry commented on libnetwork/netlabel/labels.go:
```diff
@@ -39,6 +39,14 @@ const (
+
+	// AdvertiseAddrIntervalMs is the minimum interval in milliseconds between unsolicited
+	// ARP/NA messages sent to advertise an interface's IP and MAC addresses.
+	AdvertiseAddrIntervalMs = Prefix + ".advertise_addr_ms"
```
Sure.

thaJeztah replied:
Curious / silly rquestion; is this / will this be a user-facing option to configure? In general, I agree with @corhere's comment that using Go's accepted units is non-ambiguous. I'm slightly more on the fence for some user-facing options, or at least, we've been hit by some of those because;

- users now have to make themselves familiar with Go's accepted units / formats
- and/or we end up having to replicate those in our documentation (there's some existing cases where we fell into that; in some cases for longer durations, users expected, e.g. "if I can pass minutes, hours, why can't I use `days`??")
- somewhat related; users asking "why didn't you document `µs` or `ns`?", which _technically_ was supported, but for the purpose of that option very out of range (e.g. `--health-interval`)

So, I guess "it depends" a bit;

- Is this / will this be user-facing?
- What's the expected interval we can (and are willing to) support? (e.g. is `1µs` ok, is `48h` ok?) and do we have / do we need some guardrails around that?
- _if_ there's a reasonable / expected interval for these (e.g. we'd never expect / want to support anything outside of (e.g.) `1 ms - 1000 ms` (1 second) then _possibly_ being opinionated may be more clear (guidance on what kind of range we're expecting here); it's sure is more "limited", but could be less err-prone (of course, it doesn't defeat possible need for guardrails)

robmry replied:
Yes, it's user facing - but the default 1s should basically always be fine... it's configurable up to 2s because that's what the standards allow, and down to 100ms in case that's useful (we have no experience yet). It's always limited to up-to 3 messages, so it can't cause too much damage on a real network.

I also dislike the fancy units thing, having seen the issues it causes. Units in the param name make it unambiguous for users (rather than easier for coders). But, didn't feel too strongly in this case because it's very unlikely to be widely used.

robmry replied:
@corhere - any thoughts on this?

corhere replied:
I think that units in the parameter value will result in a superior user experience than baking the unit into the parameter name. If the unit is in the name users have to remember the unit used or consult documentation to spell the parameter name correctly, with no feedback if they get it wrong. When the unit is in the value, users have fewer things to perfectly recall and we get more opportunities to catch mistakes. If the user forgets to put in a unit or otherwise provides an ill-formed duration string, we can return an error message _containing a hint that a unit suffix is required_ to save users a trip to the documentation. It could be as simple as returning the same error message as the out-of-range case!

> `error response from daemon: com.docker.network.advertise_addr_interval must be in the range 100ms to 2s`

robmry replied:
I've not made any changes here yet ... @thaJeztah - are you ok with Cory's proposal?

robmry replied:
Discussed in the maintainer's call on 9th Jan ... existing options are inconsistent, there's precedent for pretty much anything (parsed durations, seconds/milliseconds with assumed units etc). No consensus was reached, but Cory's since said he'll go with integer milliseconds for the sake of making progress here ... I've made the updates.

robmry commented on libnetwork/network_windows.go:
```diff
@@ -240,3 +241,17 @@ func defaultIpamForNetworkType(networkType string) string {
 	return defaultipam.DriverName
 }
+
+func (n *Network) validateAdvertiseAddrNMsgs() (*int, error) {
+	if _, ok := n.DriverOptions()[netlabel.AdvertiseAddrNMsgs]; ok {
+		return nil, fmt.Errorf(netlabel.AdvertiseAddrNMsgs + " is not supported on Windows")
+	}
+	return nil, nil
+}
+
+func (n *Network) validateAdvertiseAddrInterval() (*time.Duration, error) {
+	if _, ok := n.DriverOptions()[netlabel.AdvertiseAddrIntervalMs]; ok {
+		return nil, fmt.Errorf(netlabel.AdvertiseAddrIntervalMs + " is not supported on Windows")
+	}
+	return nil, nil
+}
```
The implication is you think we should silently ignore these too, how would that be an improvement?

(The issue is that we silently ignore unknown labels, and we need to deal with that. They're arcane and easy to misremember/mistype. That issue isn't this PR, but let's not add to the problem.)

corhere replied:
An overlay network with one these options set could not be instantiated on a Windows node unless the options are silently ignored on Windows networks.

robmry replied:
Ah, right - overlay networks, we can't have nice things. Updated.

robmry commented on libnetwork/osl/interface_linux.go:
```diff
@@ -20,15 +23,48 @@ import (
+	// advertiseAddrIntervalMsDefault is the default interval between ARP/NA messages
+	// sent when and interface is configured.
+	// One second matches RFC-5227 PROBE_MIN and the default for RetransTimer in RFC-4861.
+	advertiseAddrIntervalMsDefault = 1000
```
Ok.

robmry commented on libnetwork/osl/interface_linux.go:
```diff
@@ -276,22 +323,246 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+		if err != nil {
+			return fmt.Errorf("failed to refresh link attributes: %w", err)
+		}
+		if curMAC := link.Attrs().HardwareAddr; slices.Compare(curMAC, mac) != 0 {
```
Marvellous, but what shall we do with all those spare microseconds?!

thaJeztah commented on libnetwork/osl/interface_linux.go:
```diff
@@ -1,7 +1,9 @@
 import (
+	"bytes"
 	"context"
+	stderrors "errors"
```
It looks like the only place `pkg/errors` is used in this file is in `Namespace.setSysctls` (added in 0071832226b42dbb12bc01bf1c668dca891b2d33) and that function already mixes `fmt.Errorf` with `errors.Wrap`. Perhaps we're good to change those to use `errors` as well so that we don't have to use the special alias; https://github.com/moby/moby/blob/9e3e42971a887e1ef20ac4eb7c6e542cba547c34/libnetwork/osl/interface_linux.go#L462-L500



Alternatively, we could use the `internal/multierror`  package, which I think was created for better formatting, but not sure if we need to have that.

thaJeztah commented on libnetwork/osl/interface_linux.go:
```diff
@@ -276,22 +286,238 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+		"iface": i.dstName,
+		"ifi":   ifIndex,
+		"mac":   mac.String(),
+		"ip4":   i.Address(),
+		"ip6":   i.AddressIPv6(),
+	}))
+
+	if i.address == nil && i.addressIPv6 == nil {
```
Mostly curiosity; I notice that above (for logging) we're using `.MacAddress()`, `Address()` and `AddressIPv6()`, which all look to return a _copy_ of the field, but below we access some of those fields directly.

That made me curious if there's either concurrency to account for in this (in which case, it's even the question: should the interface be locked or otherwise), or if there was another reason to do so. I guess overall it made me wonder if things are manipulated if it's OK to directly access non-exported fields, from "outside" of the Interface type.

robmry replied:
Ah, yes - libnetwork locking. It doesn't make a lot of sense.

In this case, I think I had some vague thought about it being better to put copies into the logger, because the logger will be used in the arp/na-sending goroutine that'll normally run for a couple of seconds after this function returns.

But the fields are constant once initialised, so fields of an `Interface` aren't mutex protected. Even if a MAC/IP address did get updated in an `Interface`, the reference would probably be replaced (rather than the value edited in-place), so the copy would still be unnecessary but `Interface` would then need a mutex. Effectively, the values are immutable.  But, I can't prove it.

Because of that uncertainty, in a lot of places locks are acquired and copies are made when they're probably not needed. It'll remain confusing/arbitrary until we make time to sort it out.

So ... what to do in this case? I think it's ok as-is, but I'll make everything use copied versions - because that's bound to be safe and the copies are currently being made anyway.

thaJeztah commented on libnetwork/osl/interface_linux.go:
```diff
@@ -276,22 +286,238 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+	if err := n.InvokeFunc(func() {
+		if i.address != nil {
+			var err error
+			ua, err = l2disco.NewUnsolARP(ctx, i.Address().IP, i.MacAddress(), ifIndex)
```
I guess a bit similar question here; some discrepancy between accessing the fields  directly and obtaining a copy of their value.

robmry replied:
As above - I used `i.address` locally (because there's no locking on an `Interface` anyway), and `i.Address()` to make a copy of the value that'll outlive this function in the `l2disco` object.

So I think it's fine as-is too - but I'll use a copy in both places to remove the concern.

thaJeztah commented on libnetwork/sandbox_linux.go:
```diff
@@ -319,6 +319,12 @@ func (sb *Sandbox) populateNetworkResources(ctx context.Context, ep *Endpoint) e
 		if sysctls := ep.getSysctls(); len(sysctls) > 0 {
 			ifaceOptions = append(ifaceOptions, osl.WithSysctls(sysctls))
 		}
+		if nMsgs, ok := ep.advertiseAddrNMsgs(); ok {
+			ifaceOptions = append(ifaceOptions, osl.WithAdvertiseAddrNMsgs(nMsgs))
+		}
+		if interval, ok := ep.advertiseAddrInterval(); ok {
+			ifaceOptions = append(ifaceOptions, osl.WithAdvertiseAddrInterval(interval))
+		}
```
Both `advertiseAddrNMsgs` and `advertiseAddrInterval` appear to be wrappers for `network.XXX` (both do `ep.getNetwork()` as part of that).

Mostly wondering here; should we remove the extra indirection (fetch `ep.getNetwork()` once, then call these methods), or would that be too confusing?

I also noticed that both  `advertiseAddrNMsgs` and `advertiseAddrInterval`  perform validation of the values (but discarding the actual errors if any would be there); the same validation is already done on `Controller.NewNetwork()`, so curious if that state can be modified in between that requires these to be validated again. (and _if_ that's the case if validation should happen before we do all this, instead of those validation errors to be silently discarded).

robmry replied:
> Both `advertiseAddrNMsgs` and `advertiseAddrInterval` appear to be wrappers for `network.XXX` (both do `ep.getNetwork()` as part of that).
> 
> Mostly wondering here; should we remove the extra indirection (fetch `ep.getNetwork()` once, then call these methods), or would that be too confusing?

Ok, I've removed the indirection.

> I also noticed that both `advertiseAddrNMsgs` and `advertiseAddrInterval` perform validation of the values (but discarding the actual errors if any would be there); the same validation is already done on `Controller.NewNetwork()`, so curious if that state can be modified in between that requires these to be validated again. (and _if_ that's the case if validation should happen before we do all this, instead of those validation errors to be silently discarded).

It deals with Linux vs. Windows ... the config has already been validated for Linux, but the Sandbox will call this for Windows too. I've changed the names `validateAdvertiseAddrXXX` to `validatedAdvertiseAddrXXX`, and added comments to try to make that clear.

thaJeztah reviewed (approved):
LGTM

robmry commented:
Rebased, and added an error return to `UnsolARP.Close` (looks like I only added it to `UnsolNA.Close` before).

robmry commented:
I'll get this merged, to unblock https://github.com/moby/moby/pull/48809 for moby 28.0.

(@corhere - when you're back, if you have a chance to check this over again and something crops up, just ping me and I'll raise a new PR.)
